File: jiwon/leetcode/42. Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left_max, right_max = 0, 0
        left = 0
        right = len(height) - 1
        answer = 0
        while left < right:
            left_max, right_max = max(left_max, height[left]), max(right_max, height[right])
            if left_max <= right_max:
                answer += left_max - height[left]
                left += 1
            else:
                answer += right_max - height[right]
                right -= 1
        return answer

File: SeongYeon/ts/leetcode/42_TrappingRainWater.ts
function trap(height: number[]): number {
  let answer = 0;
  let left = 0;
  let right = height.length - 1;
  let leftMax = 0,
    rightMax = 0;

  while (left < right) {
    if (height[left] <= height[right]) {
      leftMax = Math.max(leftMax, height[left]);
      answer += leftMax - height[left];
      left++;
    } else {
      rightMax = Math.max(rightMax, height[right]);
      answer += rightMax - height[right];
      right--;
    }
  }

  return answer;
}
